enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,407 @@
+   places the pod on the node.
+1. Repeat steps 8-10 until all pods are running.
+1. Cluster deployment comes up with management components constrained to subset
+   of available CPUs.
+
+### API Extensions
+
+- We want to extend the `PerformanceProfile` API to include the addition of a
+  new `workloads` configuration under the `cpu` field.
+- The behavior of existing resources should not change with this addition.
+- New resources that make use of this new field will have the current machine
+  config generated with the additional machine config manifests.
+  - Uses the `isolated` to add the CRI-O and Kubelet configuration files to the
+    currently generated machine config.
+  - If no `isolated` and `enablePinning` is set to true, the default behavior is
+    to use the full CPUSet as if workloads were not pinned.
+
+Example change:
+
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New addition
+    workloads:
+      enablePinning: true
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Changes to NTO
+
+The NTO PerformanceProfile will be updated to support a new flag which will
+toggle the workload pinning to the `isolated` cores. The idea here being to
+simplify the approach for how customers set this configuration. With PAO being
+part of NTO now ([see here for more info](../node-tuning/pao-in-nto.md)) this
+affords us the chance to consolidate the configuration for `kubelet` and `crio`.
+
+We will modify the code path that generates the [new machine
+config](https://github.com/openshift/cluster-node-tuning-operator/blob/a780dfe07962ad07e4d50c852047ef8cf7b287da/pkg/performanceprofile/controller/performanceprofile/components/machineconfig/machineconfig.go#L91-L127)
+using the performance profile. With the new `workloads.enablePinning` flag we
+will add the configuration for `crio` and `kubelet` to the final machine config
+manifest. Then the existing code path will apply the change as normal.
+
+#### API Server Admission Hook
+
+The existing admission hook has 4 checks when it comes to workload pinning.
+
+1. Check if `pod` is a static pod
+   - Skips modification attempt if it is static.
+1. Checks if currently running cluster topology is Single Node
+   - Skips modification if it is anything other than Single Node
+1. Checks if all running nodes are managed
+   - Skips modification if any of the nodes are not managed
+1. Checks what resource limits and requests are set on the pod
+   - Skips modification if QoS is guaranteed or both limits and requests are set
+   - Skips modification if after update the QoS is changed
+
+We will need to alter the code in the admission controller to remove the check
+for Single Node Topology, the other configurations should remain untouched.
+
+### Risks and Mitigations
```

Comment:
Is it possible to have different worker pools with different partitioning setups? Is that desirable?

Reply:
We had a brief discussion with @browsell and @mrunalp about this. We know it is desired to have this feature turned on cluster wide and support workers and masters with different partitioning setups. So ideally we would support different partitioning setups for different worker pools. An example where this might desired is for situations where a worker pool is not desired to be partitioned, so it is setup to still be partitioned, but the CPU set is for the whole node.

Reply:
We will want to support the ability to have different partition sizes. The performancepfofile CR is per mcp so we should get this by default.

Reply:
That's right, some pools will have different partitioning settings. The existing code requires that the user defines isolated and reserved CPUSets, so it will error if none are provided. If they don't wish to apply pinning to that pool of machines, then they would set the reserved to the same as the isolated. That would allow the management workloads to have access to the full CPUSet, so technically pinning is on, but they are not limited to a smaller set of CPUs.

Reply:
OK, that feels like it could be prone to misconfiguration, but not any more so than now and using the API should make it easier to get right or fix.